src/algebra/add_submonoid_closure.lean

```diff
@@ -0,0 +1,84 @@
+/-
+Copyright (c) 2021 Damiano Testa. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Damiano Testa
+-/
+import ring_theory.subsemiring
+
+/-!
+#  Equality between `add_submonoid.closure S` and `subsemiring.closure S`
+
+Let `R` be a semiring and let `S` be a multiplicative submonoid of `R`.
+The additive closure of `S` is a subsemiring of `R`.  If `R` is commutative, then so
+is the additive closure of `S`.
+-/
+variables {R : Type*} {a b : R}
+
+section semiring
+
+variables [semiring R] {S : submonoid R}
+
+/-- The product of an element of the additive closure of a multiplicative submonoid `S`
+and an element of `S` is contained in the additive closure of `S`. -/
+lemma add_submonoid.closure_mem_mul (ha : a ∈ add_submonoid.closure (S : set R)) (hb : b ∈ S) :
+  a * b ∈ add_submonoid.closure (S : set R) :=
+begin
+  revert b,
+  refine add_submonoid.closure_induction ha _ _ _; clear ha a,
+  { exact λ r hr b hb, add_submonoid.mem_closure.mpr (λ y hy, hy (S.mul_mem hr hb)) },
+  { exact λ b hb, by simp only [zero_mul, (add_submonoid.closure (S : set R)).zero_mem] },
+  { simp_rw add_mul,
+    exact λ r s hr hs b hb, (add_submonoid.closure (S : set R)).add_mem (hr hb) (hs hb) }
+end
+
+/-- The product of two elements of the additive closure of a submonoid `S` is an element of the
+additive closure of `S`. -/
+lemma mul_mem
+  (ha : a ∈ add_submonoid.closure (S : set R)) (hb : b ∈ add_submonoid.closure (S : set R)) :
+  a * b ∈ add_submonoid.closure (S : set R) :=
+begin
+  revert a,
+  refine add_submonoid.closure_induction hb _ _ _; clear hb b,
+  { exact λ r hr b hb, add_submonoid.closure_mem_mul hb hr },
+  { exact λ b hb, by simp only [mul_zero, (add_submonoid.closure (S : set R)).zero_mem] },
+  { simp_rw mul_add,
+    exact λ r s hr hs b hb, (add_submonoid.closure (S : set R)).add_mem (hr hb) (hs hb) }
+end
+
+/-- The additive closure of a submonoid is a subsemiring. -/
+def submonoid.to_subsemiring (S : submonoid R) : subsemiring R :=
+{ carrier := add_submonoid.closure (S : set R),
+  one_mem' := add_submonoid.mem_closure.mpr (λ y hy, hy S.one_mem),
+  mul_mem' := λ x y, mul_mem,
+  zero_mem' := (add_submonoid.closure (S : set R)).zero_mem,
+  add_mem' := λ x y, (add_submonoid.closure (S : set R)).add_mem }
+
+/-- The additive closure of a submonoid is a semiring. -/
+def submonoid.to_semiring (S : submonoid R) : semiring (add_submonoid.closure (S : set R)) :=
+S.to_subsemiring.to_semiring
+
+/-- The elements of the additive closure of a multiplicative submonoid `S` are exactly the
+elements of the subsemiring closure of `S`. -/
+lemma clss : (add_submonoid.closure (S : set R)).carrier = subsemiring.closure (S : set R) :=
```

Comment:
This lemma basically already exists as `subsemiring.mem_closure_iff`

Reply:
To elaborate on my first bullet:
```suggestion
lemma clss : (add_submonoid.closure (S : set R) : set R) = subsemiring.closure (S : set R) :=
```

Reply:
Ok, I accepted your change and discovered that indeed the lemma can be proven by

`by ext; simp [subsemiring.mem_closure_iff]`!

Reply:
I think you might have my suggestion in reverse; I'm suggesting you use the machinery you've build up here to simplify the proof of `subsemiring.mem_closure_iff`!

Reply:
I see your point, but the file where `subsemiring.mem_closure_iff` is proven is also the file that introduces `subsemiring`.  It creates a circular import if I tried to prove there `subsemiring.mem_closure_iff` using the results from the file in this PR.

Reply:
An alternative would be to copy the merge this file into the `subsemiring` file, if this seems useful.

Reply:
I would probably lean towards just putting the changes in this PR straight into that file; unless you find that you can move `subsemiring.mem_closure_iff` into your new file instead

Reply:
Ok, I will try this, but likely not today!

Reply:
I finially got a chance to try this.  I moved all the content of the file `src/algebra/add_submonoid_closure` into the file `subsemiring` and managed to make the proof of `mem_closure_iff` use the new stuff.

I hope that this is what you had in mind!